File: Assets/Bundles/Client/Resources/TS/Controllers/Global/Character/LocalEntityController.lua
-- Compiled with unity-ts v2.1.0-75
local _flamework_core = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init")
local Reflect = _flamework_core.Reflect
local Flamework = _flamework_core.Flamework
local Controller = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Controller
local Game = require("Shared/TS/Game").Game
local Network = require("Shared/TS/Network").Network
local Keyboard = require("Shared/TS/UserInput/init").Keyboard
local Bin = require("Shared/TS/Util/Bin").Bin
local Signal = require("Shared/TS/Util/Signal").Signal
local Task = require("Shared/TS/Util/Task").Task
local FlyCameraMode = require("Client/TS/Controllers/Global/Camera/DefaultCameraModes/FlyCameraMode").FlyCameraMode
local HumanoidCameraMode = require("Client/TS/Controllers/Global/Camera/DefaultCameraModes/HumanoidCameraMode").HumanoidCameraMode
local FirstPersonCameraSystem = require("Client/TS/Controllers/Global/Camera/FirstPersonCameraSystem").FirstPersonCameraSystem
local EntityInput = require("Client/TS/Controllers/Global/Character/EntityInput").EntityInput
local CAM_Y_OFFSET = 1.85
local CAM_Y_OFFSET_CROUCH_1ST_PERSON = CAM_Y_OFFSET / 1.5
local CAM_Y_OFFSET_CROUCH_3RD_PERSON = CAM_Y_OFFSET
local LocalEntityController
do
	LocalEntityController = setmetatable({}, {
		__tostring = function()
			return "LocalEntityController"
		end,
	})
	LocalEntityController.__index = LocalEntityController
	function LocalEntityController.new(...)
		local self = setmetatable({}, LocalEntityController)
		return self:constructor(...) or self
	end
	function LocalEntityController:constructor(cameraController, clientSettings)
		self.cameraController = cameraController
		self.clientSettings = clientSettings
		self.firstPerson = true
		self.lookBackwards = false
		self.FirstPersonChanged = Signal.new()
		self.LookBackwardsChanged = Signal.new()
		self.customDataQueue = {}
		self.prevState = 0
		self.currentState = 0
	end
	function LocalEntityController:IsFirstPerson()
		return self.firstPerson
	end
	function LocalEntityController:ObserveFirstPerson(observer)
		local currentCleanup
		local onChanged = function(isFirstPerson)
			local _result = currentCleanup
			if _result ~= nil then
				_result()
			end
			currentCleanup = observer(isFirstPerson)
		end
		local disconnect = self.FirstPersonChanged:Connect(onChanged)
		onChanged(self.firstPerson)
		return function()
			disconnect()
			local _result = currentCleanup
			if _result ~= nil then
				_result()
			end
		end
	end
	function LocalEntityController:ObserveLookBackwards(observer)
		local currentCleanup
		local onChanged = function(lookBackwards)
			local _result = currentCleanup
			if _result ~= nil then
				_result()
			end
			currentCleanup = observer(lookBackwards)
		end
		local disconnect = self.LookBackwardsChanged:Connect(onChanged)
		onChanged(self.lookBackwards)
		return function()
			disconnect()
			local _result = currentCleanup
			if _result ~= nil then
				_result()
			end
		end
	end
	function LocalEntityController:AddToMoveData(key, value)
		local _customDataQueue = self.customDataQueue
		local _arg0 = {
			key = key,
			value = value,
		}
		table.insert(_customDataQueue, _arg0)
		local blob = BinaryBlob.new(self.customDataQueue)
		local _result = self.entityDriver
		if _result ~= nil then
			_result:SetCustomData(blob)
		end
	end
	function LocalEntityController:TakeScreenshot()
		local screenshotFilename = os.date("Screenshot-%Y-%m-%d-%H-%M-%S.png")
		print("Capturing screenshot " .. screenshotFilename)
		ScreenCapture:CaptureScreenshot(screenshotFilename)
	end
	function LocalEntityController:OnStart()
		Game.LocalPlayer:ObserveCharacter(function(entity)
			if not entity then
				return nil
			end
			local bin = Bin.new()
			local keyboard = bin:Add(Keyboard.new())
			self.entityDriver = entity.gameObject:GetComponent("EntityDriver")
			self.entityInput = EntityInput.new(entity.gameObject)
			local humanoidCameraMode
			-- Custom move data control:
			-- bin.Add(
			-- OnUpdate.ConnectWithPriority(SignalPriority.MONITOR, () => {
			-- print("LocalEntityController.ts update");
			-- if (this.customDataQueue.size() === 0) return;
			-- const blob = new BinaryBlob(this.customDataQueue);
			-- print("[TS]: SetCustomData()");
			-- entityDriver.SetCustomData(blob);
			-- }),
			-- );
			self.entityDriver:OnCustomDataFlushed(function()
				table.clear(self.customDataQueue)
			end)
			local getCamYOffset = function(state, isFirstPerson)
				local yOffset = if state == 5 or state == 4 then if isFirstPerson then CAM_Y_OFFSET_CROUCH_1ST_PERSON else CAM_Y_OFFSET_CROUCH_3RD_PERSON else CAM_Y_OFFSET
				return yOffset
			end
			-- Set up camera
			local createHumanoidCameraMode = function()
				local _result = self.entityDriver
				if _result ~= nil then
					_result = _result:GetState()
				end
				local _condition = _result
				if _condition == nil then
					_condition = 0
				end
				local state = _condition
				local yOffset = getCamYOffset(state, self.firstPerson)
				humanoidCameraMode = HumanoidCameraMode.new(entity.gameObject, entity.model, self.firstPerson, yOffset)
				humanoidCameraMode:SetLookBackwards(self.lookBackwards)
				return humanoidCameraMode
			end
			self.FirstPersonChanged:Connect(function(isFirstPerson)
				local _fn = humanoidCameraMode
				local _result = self.entityDriver
				if _result ~= nil then
					_result = _result:GetState()
				end
				local _condition = _result
				if _condition == nil then
					_condition = 0
				end
				_fn:SetYOffset(getCamYOffset(_condition, isFirstPerson), true)
			end)
			-- Set up first person camera
			self.fps = FirstPersonCameraSystem.new(entity.references, entity.dynamicVariables)
			self.fps:OnFirstPersonChanged(self.firstPerson)
			self.cameraController:SetMode(createHumanoidCameraMode())
			self.cameraController.cameraSystem:SetOnClearCallback(createHumanoidCameraMode)
			-- this.entityDriver.OnSecondaryStateChanged((state) => {
			-- humanoidCameraMode.SetYOffset(getCamYOffset(state, this.firstPerson));
			-- });
			self.entityDriver:OnStateChanged(function(state)
				if state ~= self.currentState then
					self.prevState = self.currentState
					self.currentState = state
				end
				humanoidCameraMode:SetYOffset(getCamYOffset(state, self.firstPerson))
				self:UpdateFov()
			end)
			local flyCam = false
			local flyingBin = Bin.new()
			bin:Connect(keyboard.KeyDown, function(event)
				-- Toggle first-person view:
				if event.Key == 34 then
					if self.cameraController.cameraSystem:GetMode() == humanoidCameraMode then
						self:ToggleFirstPerson(humanoidCameraMode)
					end
				elseif event.Key == 53 then
					if self.cameraController.cameraSystem:GetMode() == humanoidCameraMode then
						self:SetLookBackwards(humanoidCameraMode, true)
					end
				elseif event.Key == 30 and keyboard:IsKeyDown(51) then
					if RunCore:IsEditor() then
						flyCam = not flyCam
						if flyCam then
							if self.entityInput then
								flyingBin:Add(self.entityInput:AddDisabler())
							end
							self.cameraController:SetMode(FlyCameraMode.new())
						else
							flyingBin:Clean()
							self.cameraController:ClearMode()
						end
					end
				elseif event.Key == 27 and keyboard:IsKeyDown(51) then
					self:TakeScreenshot()
				elseif event.Key == 26 then
					print("-----")
					for _, entity in (Flamework.resolveDependency("Bundles/Client/Controllers/Global/Entity/EntityController@EntityController")):GetEntities() do
						print(entity:GetDisplayName() .. ": " .. tostring(entity.id))
					end
					print("-----")
					-- TEST: Knock-back:
					Task:Spawn(function()
						local sentTick = InstanceFinder.TimeManager.Tick
						local halfWay = Network.ClientToServer.TEST_LATENCY.Client:FireServer()
						local endTick = InstanceFinder.TimeManager.Tick
						print("Round trip: " .. tostring((endTick - sentTick)) .. " | trip 1: " .. tostring((halfWay - sentTick)) .. " | trip 2: " .. tostring((endTick - halfWay)))
					end)
				end
			end)
			bin:Connect(keyboard.KeyUp, function(event)
				if event.Key == 53 then
					if self.cameraController.cameraSystem:GetMode() == humanoidCameraMode then
						self:SetLookBackwards(humanoidCameraMode, false)
					end
				end
			end)
			-- Cleanup:
			bin:Add(function()
				self.cameraController.cameraSystem:SetOnClearCallback(nil)
				self.cameraController:ClearMode()
				local _result = self.fps
				if _result ~= nil then
					_result:Destroy()
				end
				local _result_1 = self.entityInput
				if _result_1 ~= nil then
					_result_1:Destroy()
				end
			end)
			return function()
				bin:Clean()
			end
		end)
	end
	function LocalEntityController:UpdateFov()
		local baseFov = if self:IsFirstPerson() then self.clientSettings:GetFirstPersonFov() else self.clientSettings:GetThirdPersonFov()
		if self.currentState == 3 or (self.currentState == 4 or (self.currentState == 2 and self.prevState == 3)) then
			self.cameraController:SetFOV(baseFov * 1.08, false)
		else
			self.cameraController:SetFOV(baseFov, false)
		end
	end
	function LocalEntityController:SetLookBackwards(humanoidCameraMode, lookBackwards)
		if self.lookBackwards == lookBackwards then
			return nil
		end
		self.lookBackwards = lookBackwards
		self.LookBackwardsChanged:Fire(self.lookBackwards)
		if self.cameraController.cameraSystem:GetMode() == humanoidCameraMode then
			humanoidCameraMode:SetLookBackwards(self.lookBackwards)
		end
	end
	function LocalEntityController:ToggleFirstPerson(humanoidCameraMode)
		self.firstPerson = not self.firstPerson
		self.FirstPersonChanged:Fire(self.firstPerson)
		if self.cameraController.cameraSystem:GetMode() == humanoidCameraMode then
			humanoidCameraMode:SetFirstPerson(self.firstPerson)
		end
		local _result = self.fps
		if _result ~= nil then
			_result:OnFirstPersonChanged(self.firstPerson)
		end
	end
	function LocalEntityController:GetEntityInput()
		return self.entityInput
	end
end
-- (Flamework) LocalEntityController metadata
Reflect.defineMetadata(LocalEntityController, "identifier", "Bundles/Client/Controllers/Global/Character/LocalEntityController@LocalEntityController")
Reflect.defineMetadata(LocalEntityController, "flamework:parameters", { "Bundles/Client/Controllers/Global/Camera/CameraController@CameraController", "Bundles/Client/Controllers/Global/ClientSettings/ClientSettingsController@ClientSettingsController" })
Reflect.defineMetadata(LocalEntityController, "flamework:implements", { "$:flamework@OnStart" })
Reflect.decorate(LocalEntityController, "$:flamework@Controller", Controller, { {
	loadOrder = 10000,
} })
return {
	LocalEntityController = LocalEntityController,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Bundles/Imports/Core/Core~/src/Client/Controllers/Character/LocalEntityController.ts
import { Controller, Dependency, OnStart } from "@easy-games/flamework-core";
import { CoreNetwork } from "Shared/CoreNetwork";
import { Entity } from "Shared/Entity/Entity";
import { Game } from "Shared/Game";
import { Keyboard } from "Shared/UserInput";
import { Bin } from "Shared/Util/Bin";
import { ColorUtil } from "Shared/Util/ColorUtil";
import { DataStreamItems } from "Shared/Util/DataStreamTypes";
import { Signal } from "Shared/Util/Signal";
import { Task } from "Shared/Util/Task";
import { Theme } from "Shared/Util/Theme";
import { ClientSettingsController } from "../../MainMenuControllers/Settings/ClientSettingsController";
import { CameraController } from "../Camera/CameraController";
import { FlyCameraMode } from "../Camera/DefaultCameraModes/FlyCameraMode";
import { HumanoidCameraMode } from "../Camera/DefaultCameraModes/HumanoidCameraMode";
import { OrbitCameraMode } from "../Camera/DefaultCameraModes/OrbitCameraMode";
import { FirstPersonCameraSystem } from "../Camera/FirstPersonCameraSystem";
import { EntityController } from "../Entity/EntityController";
import { InventoryController } from "../Inventory/InventoryController";
import { CharacterCameraMode } from "./CharacterCameraMode";
import { EntityInput } from "./EntityInput";

const CAM_Y_OFFSET = 1.7;
const CAM_Y_OFFSET_CROUCH_1ST_PERSON = CAM_Y_OFFSET / 1.5;
const CAM_Y_OFFSET_CROUCH_3RD_PERSON = CAM_Y_OFFSET;

@Controller({
	loadOrder: 10000,
})
export class LocalEntityController implements OnStart {
	private firstPerson = true;
	private lookBackwards = false;
	private fps?: FirstPersonCameraSystem;

	/** Fires whenever the user changes their first-person state. */
	public readonly FirstPersonChanged = new Signal<[isFirstPerson: boolean]>();

	/** Fires whenever the user requests to look (or stop looking) backwards. */
	public readonly LookBackwardsChanged = new Signal<[lookBackwards: boolean]>();

	private customDataQueue: { key: keyof DataStreamItems; value: unknown }[] = [];

	private entityDriver: EntityDriver | undefined;
	private screenshot: CameraScreenshotRecorder | undefined;
	private entityInput: EntityInput | undefined;
	private prevState: EntityState = EntityState.Idle;
	private currentState: EntityState = EntityState.Idle;
	private humanoidCameraMode: HumanoidCameraMode | undefined;
	private orbitCameraMode: OrbitCameraMode | undefined;

	private characterCameraMode: CharacterCameraMode = CharacterCameraMode.LOCKED;
	private defaultFirstPerson = true;

	constructor(
		private readonly cameraController: CameraController,
		private readonly clientSettings: ClientSettingsController,
		private readonly inventoryController: InventoryController,
	) {}

	/** Returns `true` if the player is in first-person mode. */
	public IsFirstPerson() {
		return this.firstPerson;
	}

	/** Observes the current first-person state. */
	public ObserveFirstPerson(observer: (isFirstPerson: boolean) => CleanupFunc): () => void {
		let currentCleanup: CleanupFunc;

		const onChanged = (isFirstPerson: boolean) => {
			currentCleanup?.();
			currentCleanup = observer(isFirstPerson);
			this.UpdateFov();
		};

		const disconnect = this.FirstPersonChanged.Connect(onChanged);
		onChanged(this.firstPerson);

		return () => {
			disconnect();
			currentCleanup?.();
		};
	}

	/** Observes whether or not the player wants to look backwards. */
	public ObserveLookBackwards(observer: (lookBackwards: boolean) => CleanupFunc): () => void {
		let currentCleanup: CleanupFunc;

		const onChanged = (lookBackwards: boolean) => {
			currentCleanup?.();
			currentCleanup = observer(lookBackwards);
		};

		const disconnect = this.LookBackwardsChanged.Connect(onChanged);
		onChanged(this.lookBackwards);

		return () => {
			disconnect();
			currentCleanup?.();
		};
	}

	/** Add custom data to the move data command stream. */
	public AddToMoveData<K extends keyof DataStreamItems, T extends DataStreamItems[K]>(key: K, value: T) {
		this.customDataQueue.push({ key, value });
		const blob = new BinaryBlob(this.customDataQueue);
		this.entityDriver?.SetCustomData(blob);
	}

	private TakeScreenshot(showUI: boolean, superSample: boolean) {
		if (!this.screenshot) {
			return;
		}
		let screenshotFilename = os.date("Screenshot-%Y-%m-%d-%H-%M-%S");
		const superSampleSize = superSample ? 4 : 1;
		print(`Capturing screenshot ${screenshotFilename}`);
		if (showUI) {
			this.screenshot.TakeScreenshot(screenshotFilename, superSampleSize);
		} else {
			this.screenshot.TakeCameraScreenshot(
				this.cameraController.cameraSystem.GetActiveCamera(),
				screenshotFilename,
				superSampleSize,
			);
		}
		Game.LocalPlayer.SendMessage(ColorUtil.ColoredText(Theme.Yellow, `Captured screenshot ${screenshotFilename}`));
	}

	private GetCamYOffset(state: EntityState, isFirstPerson: boolean) {
		const yOffset =
			state === EntityState.Crouching || state === EntityState.Sliding
				? isFirstPerson
					? CAM_Y_OFFSET_CROUCH_1ST_PERSON
					: CAM_Y_OFFSET_CROUCH_3RD_PERSON
				: CAM_Y_OFFSET;
		return yOffset;
	}

	private CreateHumanoidCameraMode(entity: Entity): HumanoidCameraMode {
		const state = this.entityDriver?.GetState() ?? EntityState.Idle;
		const yOffset = this.GetCamYOffset(state, this.firstPerson);
		this.humanoidCameraMode = new HumanoidCameraMode(entity.gameObject, entity.model, this.firstPerson, yOffset);
		this.humanoidCameraMode.SetLookBackwards(this.lookBackwards);
		return this.humanoidCameraMode;
	}

	private CreateOrbitCameraMode(entity: Entity): OrbitCameraMode {
		return new OrbitCameraMode(entity.model.transform, 6);
	}

	OnStart() {
		Game.LocalPlayer.ObserveCharacter((entity) => {
			if (!entity) return;

			const bin = new Bin();
			const keyboard = bin.Add(new Keyboard());
			this.firstPerson = this.defaultFirstPerson;

			this.entityDriver = entity.gameObject.GetComponent<EntityDriver>();
			this.entityInput = new EntityInput(entity);

			this.screenshot = entity.gameObject.AddComponent<CameraScreenshotRecorder>();

			this.entityDriver.OnCustomDataFlushed(() => {
				this.customDataQueue.clear();
			});

			// Set up camera
			if (this.characterCameraMode === CharacterCameraMode.LOCKED) {
				this.cameraController.SetMode(this.CreateHumanoidCameraMode(entity));
				this.cameraController.cameraSystem.SetOnClearCallback(() => this.CreateHumanoidCameraMode(entity));
			} else if (this.characterCameraMode === CharacterCameraMode.ORBIT) {
				this.cameraController.SetMode(this.CreateOrbitCameraMode(entity));
				this.cameraController.cameraSystem.SetOnClearCallback(() => this.CreateOrbitCameraMode(entity));
			}

			this.FirstPersonChanged.Connect((isFirstPerson) => {
				this.humanoidCameraMode?.SetYOffset(
					this.GetCamYOffset(this.entityDriver?.GetState() ?? EntityState.Idle, isFirstPerson),
					true,
				);
			});

			//Set up first person camera
			this.fps = new FirstPersonCameraSystem(entity.references, this.firstPerson);

			this.entityDriver.OnStateChanged((state) => {
				if (state !== this.currentState) {
					this.prevState = this.currentState;
					this.currentState = state;
				}
				this.humanoidCameraMode?.SetYOffset(this.GetCamYOffset(state, this.firstPerson));
				this.UpdateFov();
			});

			let flyCam = false;

			const flyingBin = new Bin();

			// Pause Editor
			keyboard.OnKeyDown(KeyCode.Tilde, (event) => {
				DebugUtil.TogglePauseEngine();
			});
			keyboard.OnKeyDown(KeyCode.BackQuote, (event) => {
				DebugUtil.TogglePauseEngine();
			});

			// Toggle first person:
			keyboard.OnKeyDown(KeyCode.T, (event) => {
				if (this.cameraController.cameraSystem.GetMode() === this.humanoidCameraMode) {
					this.ToggleFirstPerson();
				}
			});

			// Toggle fly cam:
			keyboard.OnKeyDown(KeyCode.P, (event) => {
				if (keyboard.IsKeyDown(KeyCode.LeftShift)) {
					if (flyCam) {
						flyCam = false;
						flyingBin.Clean();
					} else {
						flyCam = true;
						let backToFirstPerson = this.firstPerson;
						if (backToFirstPerson) {
							this.SetFirstPerson(false);
						}
						this.cameraController.SetMode(new FlyCameraMode());
						flyingBin.Add(() => {
							this.cameraController.ClearMode();
							if (backToFirstPerson) {
								this.SetFirstPerson(true);
							}
						});
						flyingBin.Add(this.entityInput!.AddDisabler());
						flyingBin.Add(this.inventoryController.AddDisabler());
					}
				}
			});

			// Toggle fly mode (like mc creative):
			let lastSpace = 0;
			keyboard.OnKeyDown(KeyCode.Space, (event) => {
				const now = Time.time;
				const dt = now - lastSpace;
				if (dt < 0.3) {
					lastSpace = 0;
					if (this.entityDriver?.IsAllowFlight()) {
						this.entityDriver?.SetFlying(!this.entityDriver.IsFlying());
					}
				} else {
					lastSpace = now;
				}
			});

			// Toggle look backwards:
			keyboard.OnKeyDown(KeyCode.LeftAlt, (event) => {
				if (this.cameraController.cameraSystem.GetMode() === this.humanoidCameraMode) {
					this.SetLookBackwards(true);
				}
			});
			keyboard.OnKeyUp(KeyCode.LeftAlt, (event) => {
				if (this.cameraController.cameraSystem.GetMode() === this.humanoidCameraMode) {
					this.SetLookBackwards(false);
				}
			});

			// Screenshot:
			keyboard.OnKeyDown(KeyCode.F2, (event) => {
				this.TakeScreenshot(keyboard.IsKeyDown(KeyCode.LeftShift), keyboard.IsKeyDown(KeyCode.LeftControl));
			});

			// Debug knockback:
			keyboard.OnKeyDown(KeyCode.L, (event) => {
				print("-----");
				for (const entity of Dependency<EntityController>().GetEntities()) {
					print(entity.GetDisplayName() + ": " + entity.id);
				}
				print("-----");
				// TEST: Knock-back:
				Task.Spawn(() => {
					const sentTick = InstanceFinder.TimeManager.Tick;
					const halfWay = CoreNetwork.ClientToServer.TEST_LATENCY.Client.FireServer();
					const endTick = InstanceFinder.TimeManager.Tick;
					print(
						"Round trip: " +
							(endTick - sentTick) +
							" | trip 1: " +
							(halfWay - sentTick) +
							" | trip 2: " +
							(endTick - halfWay),
					);
				});
			});

			keyboard.OnKeyDown(KeyCode.Semicolon, (event) => {
				CoreNetwork.ClientToServer.TestKnockback2.Client.FireServer();
			});

			//Libonati Test Space - DONT COMMIT
			keyboard.OnKeyDown(KeyCode.G, (event) => {
				print("Sending Libonati Debug Command");
				CoreNetwork.ClientToServer.LibonatiTest.Client.FireServer();
			});

			// Cleanup:
			bin.Add(() => {
				this.cameraController.cameraSystem.SetOnClearCallback(undefined);
				this.cameraController.ClearMode();
				this.fps?.Destroy();
				this.entityInput?.Destroy();
			});

			entity.OnDeath.Connect(() => {
				bin.Clean();
			});

			return () => {
				bin.Clean();
			};
		});
	}

	public SetCharacterCameraMode(mode: CharacterCameraMode): void {
		this.characterCameraMode = mode;

		if (Game.LocalPlayer.Character) {
			const entity = Game.LocalPlayer.Character;
			if (mode === CharacterCameraMode.LOCKED) {
				this.cameraController.SetMode(this.CreateHumanoidCameraMode(entity));
			} else if (mode === CharacterCameraMode.ORBIT) {
				this.cameraController.SetMode(this.CreateOrbitCameraMode(entity));
			}
		}
	}

	public UpdateFov(): void {
		let baseFov = this.IsFirstPerson()
			? this.clientSettings.GetFirstPersonFov()
			: this.clientSettings.GetThirdPersonFov();
		if (
			// this.currentState === EntityState.Sprinting ||
			this.currentState === EntityState.Sliding
			// (this.currentState === EntityState.Jumping && this.prevState === EntityState.Sprinting)
		) {
			this.cameraController.SetFOV(baseFov * 1.08, false);
		} else {
			this.cameraController.SetFOV(baseFov, false);
		}
	}

	private SetLookBackwards(lookBackwards: boolean) {
		if (this.lookBackwards === lookBackwards) return;
		this.lookBackwards = lookBackwards;
		this.LookBackwardsChanged.Fire(this.lookBackwards);

		if (this.cameraController.cameraSystem.GetMode() === this.humanoidCameraMode) {
			this.humanoidCameraMode.SetLookBackwards(this.lookBackwards);
		}
	}

	public ToggleFirstPerson() {
		this.SetFirstPerson(!this.firstPerson);
	}

	public SetFirstPerson(value: boolean) {
		if (this.firstPerson === value) {
			return;
		}

		this.firstPerson = value;
		this.FirstPersonChanged.Fire(this.firstPerson);

		if (this.cameraController.cameraSystem.GetMode() === this.humanoidCameraMode) {
			this.humanoidCameraMode.SetFirstPerson(this.firstPerson);
		}
		this.fps?.OnFirstPersonChanged(this.firstPerson);
	}

	public GetEntityInput(): EntityInput | undefined {
		return this.entityInput;
	}

	public SetDefaultFirstPerson(val: boolean): void {
		this.defaultFirstPerson = val;
	}

	public IsDefaultFirstPerson(): boolean {
		return this.defaultFirstPerson;
	}
}
